### PR TITLE
allow for overlapping beatlooprolls

### PR DESCRIPTION
**Pull request by iamcodemaker**
This is necessary if multiple beatlooproll keys are pressed on a controller. Previously, Mixxx would stop the loop roll if any keys were released. Now it tracks all active rolls and will activate the last active one that is still active (button still down) instead of deactivating all rolls. Additionally, if a loop roll is active, activating another beat loop roll will start the loop from the start of the current loop, instead the current play position.

I believe this loop roll behavior is more intuitive and natural.

### CODE REVIEW

**Comment by daschuer** [comment]
Thank you for the Pull request.
Do we have a Launchpad bug for it?

**Review by daschuer — CHANGES_REQUESTED** [review]
Works good, I have only some minor comments. 
Thank you.

File: src/engine/loopingcontrol.cpp [hunk 1]
```diff
@@ -846,10 +847,11 @@ void LoopingControl::slotBeatLoopActivateRoll(BeatLoopingControl* pBeatLoopContr
          return;
      }
 
-    // Disregard existing loops.
+    // Disregard existing loops (except beatlooprolls).
     m_pSlipEnabled->set(1);
-    slotBeatLoop(pBeatLoopControl->getSize(), false, true);
+    slotBeatLoop(pBeatLoopControl->getSize(), m_bLoopRollActive, true);
     m_bLoopRollActive = true;
+    m_activeLoopRolls.push(pBeatLoopControl);
```
[the remark this answers is not in the document]
Comment (iamcodemaker) [reply src/engine/loopingcontrol.cpp]: Yes. I'll make that change.

File: src/engine/loopingcontrol.cpp [hunk 2]
```diff
@@ -858,14 +860,21 @@ void LoopingControl::slotBeatLoopDeactivate(BeatLoopingControl* pBeatLoopControl
 }
 
 void LoopingControl::slotBeatLoopDeactivateRoll(BeatLoopingControl* pBeatLoopControl) {
-    Q_UNUSED(pBeatLoopControl);
-    setLoopingEnabled(false);
+    pBeatLoopControl->deactivate();
+    m_activeLoopRolls.removeAll(pBeatLoopControl);
```
[the remark this answers is not in the document]
Comment (iamcodemaker) [reply src/engine/loopingcontrol.cpp]: Will update. In theory there should only be one of each beat loop in here max. Actually thinking about it, if a controller had two ways to activate a loop roll we might end up with duplicates. So looping here is necessary.

File: src/engine/loopingcontrol.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (iamcodemaker) [reply src/engine/loopingcontrol.cpp]: If we store `double`s in here, are we going to have any issues comparing them using `==`? I don't understand floats well enough to say. The values should be exact though, so my gut says this is fine.

File: src/engine/loopingcontrol.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (iamcodemaker) [reply src/engine/loopingcontrol.cpp]: Made the removal loop O(n) (though the stack will always be so small it won't matter). It is slightly more complex now though.

**Comment by ronso0** [comment]
Ooouh, sounds interesting!! I got used to the way it works right now, but your idea makes sense. Didn't test this yet.


Is this the right place to ask for taking a look at [lp:1721881 beatloop shouldn't alter beatloop size in spinbox](https://bugs.launchpad.net/mixxx/+bug/1721881) ?


Also, just yesterday I wished I could transfrom a **rolling** beatlopp into a 'regular' one, to resize it and move it. I got stucked when thinking about the controller implementation...

**Comment by rryan** [comment]
Thanks! To make sure we dont break this in the future could you please add some unit tests?
src/test/looping_control_test.cpp

**Comment by iamcodemaker** [comment]
Not sure if there is a launchpad bug for it, I didn't check. I'll work on some tests.

**Comment by daschuer** [comment]
@rons0 transforming a loop, in a way that you can release you finger and the loop keeps looping?
Thats a valid use case. Can you file a bug that we can consider details there?

**Comment by ronso0** [comment]
[lp:1793520 transform a rolling beatloop into a 'static' loop](https://bugs.launchpad.net/mixxx/+bug/1793520)

**Review by daschuer — APPROVED** [review]

**Comment by daschuer** [comment]
The Travis failure is unrelated.
LGTM Thank you very much.

**Pull request merged**